### PR TITLE
Gene gallery handler

### PR DESCRIPTION
**Pull request by TriLe965**
- Use `ClassFactory` to create real classes from abstract classes (a design pattern suggested by Nam+Minh. Don't ask me...)
- `JSONFile` ensures read/write of encrypted json files. It sends en/decryption requests on SDKClient and en/decrypt directly on SDKServer.
- `GeneGallery` handles gene gallery information in a BBrowser
- `SDKServer` can start locally based on an abstract `LocationManager` -> If starts locally, SDKServer can read local data based on BBrowser's config
- Unit tests for `JSONFile`, `GeneGallery`, and `LocationManager`. To run the test, go to "BBrowserSDK" folder and run `python -m pytest`

TODO:
- `GeneGallery` uses ensembl IDs -> needs a nomenclature converter based on the species in run_info.json

### CODE REVIEW

File: BioTuringSDK/GeneGallery.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    def read(self):
+        if not self.exists():
+            raise Exception("No gallery data to read")
+        
+        if self.meta_file.exists():
+            self.collections = self.meta_file.read()
+            return
+        
+        # TODO: Handle old gallery data
```
Comment (JATC1024) [inline BioTuringSDK/GeneGallery.py:119]: Should raise UnimplementedError

File: BioTuringSDK/GeneGallery.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            colId (str): collection ID
+        """
+        now = helpers.get_timestamp()
+        col = GalleryCollection(name=name, type=ft_type, id=helpers.generate_uuid(), items=[],
+                                created_at=now, last_modified=now)
+        self.collections.append(col)
+        return col["id"]
+    
+    def __get_collection_index(self, col_id: str):
+        for i, col in enumerate(self.collections):
```
Comment (JATC1024) [inline BioTuringSDK/GeneGallery.py:71]: What if there is no col_id?

File: BioTuringSDK/GeneGallery.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        return col_id
+    
+    def get(self, col_id: str, straighten: bool=True, unique: bool=True) -> list[str]:
+        """Get a list of item IDs given a collection ID"""
+        col = self.collections[self.__get_collection_index(col_id)]
+        result = []
+        for item in col["items"]:
+            if straighten:
+                for ft in item["features"]:
+                    if unique and ft not in result:
```
Comment (JATC1024) [inline BioTuringSDK/GeneGallery.py:97]: Potential bug: what if unique=False?

File: BioTuringSDK/GeneGallery.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        return item["id"]
+    
+    def create_gene_collection(self, name: str, gene_id: list[str]) -> str:
+        """Create a gene collection given a list of genes"""
+        col_id = self.create_empty_collection(name)
+        for gene in gene_id:
+            self.add_item(col_id, gene, [gene]) # FIXME: Handle gene name and ID
+        return col_id
+    
+    def get(self, col_id: str, straighten: bool=True, unique: bool=True) -> list[str]:
```
Comment (JATC1024) [inline BioTuringSDK/GeneGallery.py:90]: This function does not seem to always return a list of strings. When straighten=True, it returns a union of list[str] and list[list[str]].

File: BioTuringSDK/factory.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        JSONFile = self.get_class_JSONFile()
+
+        class GeneGallery(AbstractGeneGallery):
+            def get_class_JSONFile(self):
+                return JSONFile
+        
+        return GeneGallery
+    
+    def get_class_LocationManager(self):
+        if self.endpoint:
```
Comment (JATC1024) [inline BioTuringSDK/factory.py:83]: `if self.endpoint:` returns True when endpoint is NOT None, which means this is a REMOTE server.

File: BioTuringSDK/factory.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                    if setting_file.exists():
+                        settings = setting_file.read()
+                        if "location" in settings:
+                            self.data_location = Path(settings["location"], root_name)
+                        if "module_location" in settings:
+                            self.module_location = Path(settings["module_location"], root_name)
+                
+                def get_private_dir(self, user_name: str="") -> Path:
+                    return self.data_location.joinpath("Data", "SingleCell", "Study")
+            return LocalLocationManager
```
Comment (JATC1024) [inline BioTuringSDK/factory.py:104]: so this must return `BESPLocationManager` instead?

**Comment by TriLe965** [comment]
Moved to
https://github.com/BioTuring-Notebooks/BBrowserSDK/pull/76

**Pull request closed without merging**